src/main/java/com/microsoft/sqlserver/jdbc/SQLServerCertificateUtils.java

```diff
@@ -0,0 +1,276 @@
+        keystore.load(new FileInputStream(certPath), keyPassword.toCharArray());
+        KeyManagerFactory keyManagerFactory = KeyManagerFactory.getInstance("SunX509");
+        keyManagerFactory.init(keystore, keyPassword.toCharArray());
+        return keyManagerFactory.getKeyManagers();
+    }
+
+    private static KeyManager[] readPKCS8Certificate(String certPath, String keyPath,
+            String keyPassword) throws IOException, GeneralSecurityException, SQLServerException {
+        Certificate clientCertificate = loadCertificate(certPath);
+        PrivateKey privateKey = loadPrivateKey(keyPath, keyPassword);
+
+        KeyStore keyStore = KeyStore.getInstance("JKS");
+        keyStore.load(null, null);
+        keyStore.setCertificateEntry("client-cert", clientCertificate);
+        keyStore.setKeyEntry("client-key", privateKey, keyPassword.toCharArray(),
+                new Certificate[] {clientCertificate});
+
+        KeyManagerFactory kmf = KeyManagerFactory.getInstance(KeyManagerFactory.getDefaultAlgorithm());
+        kmf.init(keyStore, keyPassword.toCharArray());
+        return kmf.getKeyManagers();
+    }
+
+    private static Certificate loadCertificate(String certificatePem) throws IOException, GeneralSecurityException {
+        CertificateFactory certificateFactory = CertificateFactory.getInstance("X509");
+        InputStream certstream = fileToStream(certificatePem);
+        return certificateFactory.generateCertificate(certstream);
+    }
+
+    // PKCS#8 format
+    private static final String PEM_PRIVATE_START = "-----BEGIN PRIVATE KEY-----";
+    private static final String PEM_PRIVATE_END = "-----END PRIVATE KEY-----";
+    // PKCS#1 format
+    private static final String PEM_RSA_PRIVATE_START = "-----BEGIN RSA PRIVATE KEY-----";
+    // PVK format
+    private static final long PVK_MAGIC = 0xB0B5F11EL;
+    private static final byte[] RSA2_MAGIC = {82, 83, 65, 50};
+    private static final String RC4_ALG = "RC4";
+    private static final String RSA_ALG = "RSA";
+
+    private static PrivateKey loadPrivateKey(String privateKeyPemPath,
+            String privateKeyPassword) throws GeneralSecurityException, IOException, SQLServerException {
+        String privateKeyPem = getStringFromFile(privateKeyPemPath);
+
+        if (privateKeyPem.contains(PEM_PRIVATE_START)) { // PKCS#8 format
+            return loadPrivateKeyFromPKCS8(privateKeyPem);
+        } else if (privateKeyPem.contains(PEM_RSA_PRIVATE_START)) { // PKCS#1 format
+            return loadPrivateKeyFromPKCS1(privateKeyPem, privateKeyPassword);
+        } else {
+            return loadPrivateKeyFromPVK(privateKeyPemPath, privateKeyPassword);
+        }
+    }
+
+    private static PrivateKey loadPrivateKeyFromPKCS8(
+            String key) throws IOException, NoSuchAlgorithmException, InvalidKeySpecException {
+        key = key.replace(PEM_PRIVATE_START, "").replace(PEM_PRIVATE_END, "");

```

Comment:
Is it possible to use the StringBuilder to do these manipulations on the key? 
The current code will create new instances of String since String is immutable. Can we optimize a bit for memory?